File: backend/app/api/router/subscriptions.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.controller.subscriptions import SubscriptionController
from app.authen.current_user import AuthUser, is_admin_user, required_roles
from app.db.session import get_db
from app.enums.parking import PaymentType, SubscriptionPlanType, SubscriptionStatus
from app.models.responses import DeleteResponse
from app.models.subscriptions import (
    UserSubscriptionCreate,
    UserSubscriptionClientView,
    UserSubscriptionAdminView,
    UserSubscriptionRead,
)
from app.utils.pagination import PaginatedResponse

router = APIRouter(prefix="/subscriptions", tags=["user_subscriptions"])


@router.post("/", response_model=UserSubscriptionRead)
async def create_subscription(payload: UserSubscriptionCreate, db: AsyncSession = Depends(get_db)):
    return await SubscriptionController.create_subscription_ctrl(payload, db)

@router.get("/me/paginated", response_model=PaginatedResponse[UserSubscriptionClientView])
async def get_current_user_subscriptions_paginated(
    current_user: AuthUser = Depends(required_roles("USER")),
    db: AsyncSession = Depends(get_db),
    status: SubscriptionStatus | None = Query(None, description="Filter by subscription status"),
    page: int = Query(1, ge=1, description="Page number (1-based)"),
    limit: int = Query(5, ge=1, le=100, description="Number of items per page"),
):
    return await SubscriptionController.get_user_subscriptions_by_user_paginated_ctrl(
        current_user.user_code,
        db,
        status=status,
        page=page,
        limit=limit,
    )

@router.get("/me", response_model=list[UserSubscriptionAdminView])
async def get_current_user_subscriptions(
    current_user: AuthUser = Depends(required_roles("USER")),
    db: AsyncSession = Depends(get_db),
    status: SubscriptionStatus | None = Query(
        None,
        description="Filter by one subscription status",
    ),
    statuses: list[SubscriptionStatus] | None = Query(
        None,
        description="Filter by multiple subscription statuses",
    ),
):
    return await SubscriptionController.get_user_subscriptions_by_user_ctrl(
        current_user.user_code,
        db,
        status=status,
        statuses=statuses,
    )

@router.get("/details/paginated", response_model=PaginatedResponse[UserSubscriptionAdminView])
async def get_subscription_details_paginated(
    current_user: AuthUser = Depends(required_roles("ADMIN")),
    db: AsyncSession = Depends(get_db),
    search: str | None = Query(None, description="Legacy search (user, plan, plate, term)"),
    user_code: str | None = Query(None, description="Search by user code"),
    full_name: str | None = Query(None, description="Search by full name"),
    plan_type: SubscriptionPlanType | None = Query(None, description="Filter by plan type"),
    payment_type: PaymentType | None = Query(None, description="Filter by payment plan"),
    status: SubscriptionStatus | None = Query(None, description="Filter by subscription status"),
    page: int = Query(1, ge=1, description="Page number (1-based)"),
    limit: int = Query(5, ge=1, le=100, description="Number of items per page"),
):
    return await SubscriptionController.get_all_subscription_details_paginated_ctrl(
        db,
        search=search,
        user_code=user_code,
        full_name=full_name,
        plan_type=plan_type,
        payment_type=payment_type,
        status=status,
        page=page,
        limit=limit,
    )

@router.delete("/{subscription_id}", response_model=DeleteResponse)
async def delete_subscription(subscription_id: str, db: AsyncSession = Depends(get_db)):
    return await SubscriptionController.delete_subscription_ctrl(subscription_id, db)

File: dashboard/src/api/subscriptions.ts
import { useQuery } from '@tanstack/react-query';

import { httpGet } from './httpClient';
import type { PaginatedResponse, UserSubscriptionDetailRecord } from './types';
import type { subscriptionStatusOptions, planTypeOptions } from '../constant/config';

type SubscriptionDetailsQuery = {
  search?: string;
  user_code?: string;
  full_name?: string;
  plan_type?: typeof planTypeOptions[keyof typeof planTypeOptions];
  payment_type?: 'FULL' | 'MONTHLY';
  status?: typeof subscriptionStatusOptions[keyof typeof subscriptionStatusOptions];
  page: number;
  limit: number;
};

const fetchSubscriptionDetailsPaginated = async (params: SubscriptionDetailsQuery) => {
  const search = new URLSearchParams();
  if (params.search) search.append('search', params.search);
  if (params.user_code) search.append('user_code', params.user_code);
  if (params.full_name) search.append('full_name', params.full_name);
  if (params.plan_type) search.append('plan_type', params.plan_type);
  if (params.payment_type) search.append('payment_type', params.payment_type);
  if (params.status) search.append('status', params.status);
  search.append('page', String(params.page));
  search.append('limit', String(params.limit));
  const query = search.toString();
  return httpGet<PaginatedResponse<UserSubscriptionDetailRecord>>(
    `/subscriptions/details/paginated${query ? `?${query}` : ''}`
  );
};

export const useSubscriptionDetailsPaginated = (params: SubscriptionDetailsQuery) => {
  return useQuery({
    queryKey: ['admin', 'subscriptionDetails', 'paginated', params],
    queryFn: () => fetchSubscriptionDetailsPaginated(params),
    staleTime: 1000 * 30,
  });
};
